---
title: "PseudoBulk V2"
author: "Brian Gural"
date: "2023-08-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load libs, message=FALSE, echo = F, warning=FALSE, results= 'hide', cache=FALSE, include=F}
# List libraries
libs <- c("Seurat", "ggplot2", "patchwork", "SeuratDisk","reshape2", "tidyverse",
          "SCpubr","shiny", "ggrepel", "gridExtra", "scCustomize", "httr", 
          "scales", "dplyr", "DeconvoBuddies", "readr", "SingleCellExperiment",
          "SummarizedExperiment", "Biobase", "ggmagnify", "stringr", "MuSiC",
          "ggforce", "DirichletReg", "readxl", "viridis", "DESeq2", "rlist",
          "purrr") # list libraries here
# Require all of them
lapply(libs, require, character.only = T)

rm(libs)

```

```{r load data, cache = TRUE, echo=FALSE, message=F, include = F}
sn <- LoadH5Seurat("data/processed/single_cell/merges/rau_patterson/09132023/cell_types_db_1_5.h5seurat")
```

```{r make downsampled df}
# Downsample 
sn.small <- sn |>
  GetAssayData(assay = "RNA", slot = "counts") |>
  as.matrix() |>
  SampleUMI(4000, verbose = T, upsample = T)

sn.small <- CreateSeuratObject(sn.small)
meta.features <- colnames(sn@meta.data)
for(i in meta.features){
  sn.small <- AddMetaData(sn.small, sn@meta.data[[i]], col.name = i)
}

Idents(sn.small) <- Idents(sn)

library(Seurat)
library(stringi)

# Replace spaces, commas, and periods in Idents levels with underscores
new_levels <- gsub("[,\\. ]", "_", levels(Idents(sn.small)))

# Update Idents with the new levels
Idents(sn.small) <- factor(Idents(sn.small), levels = levels(Idents(sn.small)), labels = new_levels)

# Verify the new levels
levels(Idents(sn.small))

```
## R Markdown

Having highly correlated variables in the design leads to unstable estimates of coefficients which will be reflected in large standard errors - Mike Love

This Rmd is meant to test the effect of different representations of composition in DESeq2 models. I've made simulated bulk RNAseq datasets of from range of cell type proportions, using snRNAseq as a gene expression profile reference. The null hypothesis is that compositional differences will not result in false positives between samples. The alternative hypothesis is that there will be differences and the can be accounted for with covariates which represent the composition.


```{r make pb functions, include = F}

# Function to make a data frame of cell type ratios
simulate_ratios <- function(major_cell, changing_cell, major_prop, changing_prop, 
                            cell_types, range, step_size, replicates, noise) {
  
  # get group tiers for major cell type
  # these are the percentages of the major cell type for which groups will be centered
  major_props_groups <- seq(major_prop - range, major_prop + range, step_size)
  
  # get values for major cell type
  # each value in the major_props_groups vector is used as a mean with n replicates pulled from a normal distribution with noise
  major_props <- lapply(major_props_groups, function(x){rnorm(replicates, x, noise)}) |>
                   unlist()
  
  # get group tiers for changing cell type
  changing_props_groups <- rev(seq(changing_prop - range, changing_prop + range, step_size))
  
  #get values for changing cell type
  changing_props <- lapply(changing_props_groups, function(x){rnorm(replicates, x, noise)}) |>
                      unlist()
  
  # Set major cell type and proportion
  minor <- cell_types[!(cell_types %in% c(major_cell, changing_cell))]
  
  
  # Define the range for proportions
  minor_props <- rep((1-major_props-changing_props)/length(minor), length(minor)) |>
                 matrix(ncol = length(minor)) |>
                 as.data.frame()
  colnames(minor_props) <- minor
  
  # set up props dataframe
  combos <- c(major_props, changing_props) |>
            matrix(ncol = 2) |>
            as.data.frame()
  colnames(combos) <- c(major_cell, changing_cell)
  
  # add minor cell types to props data frame
  props <- cbind(combos, minor_props)
  props$pct.change <- as.factor(rep(changing_props_groups - changing_prop, each = replicates))
  props$pct.change <- relevel(props$pct.change, ref = "0")
  rownames(props) <- paste0("mix_", 1:nrow(props))
  
  # add 3rd and 4th cell type proportions with a very small amount of noise
  props[,3] <- props[,3] * sample(rnorm(1000, 1, 0.0005), length(changing_props))
  props[,4] <- props[,4] * sample(rnorm(1000, 1, 0.0003), length(changing_props))
  # Normalize back to proportional sum == 1
  props[,1:length(cell_types)]  <- props[,1:length(cell_types)]/rowSums(props[,1:length(cell_types)])
  
  return(props)
}


# Function to sample cell IDs from a cell type, then aggregate their expression
AddCellProfile <- function(sn, count, cell_type){
    
    # Get the cell.ids that are the cell type of interest
    cell.ids  <- Idents(sn)[which(Idents(sn) == cell_type)] |>
              names() |>
              sample(size = count[cell_type], replace = T)
     
     # sum expression of cells with those ids
     cell.type.profile <- rowSums(sn@assays$RNA@counts[,cell.ids]) |>
       as.numeric()  
     assign(as.character(cell_type), cell.type.profile) # name the profile with the cell type
    return(get(as.character(cell_type)))
  }
  
AggCells <- function(sn, ratios, cell_count) {
  ratio.format <- ratios[1,] |> unlist()
  names(ratio.format) <- names(ratios)
  # Ensure ratios is a named list or named vector
  if (!is.list(ratio.format) && !is.vector(ratio.format)) {
    stop("ratios must be a named list or named vector")
  }
  # Cell counts by type
  cell_counts <- round(ratio.format * cell_count)
  #names(cell_counts) <- seq(0,length(ratio.format)-1, 1)
  dat <- sapply(1:length(cell_counts), function(x)
            {
            AddCellProfile(sn, cell_counts[x], names(cell_counts)[x])
            }
               ) |>
            rowSums()
  names(dat) <- rownames(sn)
  return(dat)
}


```


```{r make pb ratios}

major.cell <- "Cardiomyocytes"
changing.cell <- "Fibroblast__Mesen_"
major.prop <- 0.3
changing.prop <- 0.4
cell.types <- unfactor(unique(Idents(sn.small)))
range <- 0.2
step.size <- 0.01
replicates <- 10
noise <- 0.005

ratios <- simulate_ratios(major_cell = major.cell, 
                                  changing_cell = changing.cell, 
                                  major_prop = major.prop, 
                                  changing_prop = changing.prop, 
                                  cell_types = cell.types, 
                                  range = range, 
                                  step_size = step.size, 
                                  replicates = replicates,
                                  noise = noise)

```

```{r make additive pseudobulk}
start <- Sys.time()
# Function to make a list of cell type ratios
cell.target <- 5000
# take each row of cell.ratios and put it into MakePseudoBulk 
pb.add <- sapply(1:nrow(ratios), function(x){AggCells(sn.small, ratios[x,-10], cell.target)}) |>
      as.data.frame()
colnames(pb.add) <- rownames(ratios)

end <- Sys.time() - start
```


```{r comp vars prep}
# Make clr comp
library(compositions)
clr.sample <- clr(ratios[,as.character(cell.types)])
colnames(clr.sample) <- paste0("clr.", colnames(clr.sample))

# Make PCA comp
pca.sample <- prcomp(t(ratios[,as.character(cell.types)]))$rotation |>
              as.data.frame()

pca.sample$pct.change <- ratios$pct.change
pca.sample$fb <- ratios[changing.cell]
pca.sample$cm <- ratios[major.cell]

# add this info back to proportion dfs 
ratios <- cbind(ratios, clr.sample) |> 
                 cbind(pca.sample) 
```


```{r make deseq2 df }
# make df with clr, pca, and raw compositions
counts <- pb.add
counts.small <- counts[sample(rownames(counts), 200),]
```

```{r induced expression changes}

# Randomly select 10% of genes to be differentially expressed
n_genes <- nrow(counts.small)
n_diff_exp_genes <- round(0.2 * n_genes)
diff_exp_genes <- sample(rownames(counts.small), n_diff_exp_genes)
stable_genes <- rownames(counts.small)[!(rownames(counts.small) %in% diff_exp_genes)]
# Define the fold-change or other modifier
fold_change <- 2  # Change this based on your specific needs
reference.group <- ratios |> 
                    subset(pct.change == 0) %>%
                    rownames_to_column() |> 
                    pull(rowname)
# Loop over the columns (samples) in counts.small to apply the fold-change
# Skip the reference group column(s) as needed
for (col in colnames(counts.small)) {
  if (col %in% reference.group) {
    next
  }
  counts.small[diff_exp_genes, col] <- counts.small[diff_exp_genes, col] * fold_change
}

```

```{r deseq2 model comparison improvement}

models <- list(
  unadjusted    = ~ 0 + pct.change,
  raw_cm        = ~ 0 + pct.change + Cardiomyocytes,
  raw_cm_fb     = ~ 0 + pct.change + Cardiomyocytes + Fibroblast__Mesen_,
  raw_cm_fb_im  = ~ 0 + pct.change + Cardiomyocytes + Fibroblast__Mesen_ + Endothelial__Cor__Art_,
  clr_cm        = ~ 0 + pct.change + clr.Cardiomyocytes,
  clr_cm_fb     = ~ 0 + pct.change + clr.Cardiomyocytes + clr.Fibroblast__Mesen_,
  clr_cm_fb_im  = ~ 0 + pct.change + clr.Cardiomyocytes + clr.Fibroblast__Mesen_ + clr.Endothelial__Cor__Art_,
  pc1           = ~ 0 + pct.change + PC1,
  pc2           = ~ 0 + pct.change + PC1 + PC2
  )


```

```{r old deseq function, include = F}

sample_info <- ratios[colnames(counts),]
colnames(sample_info) <- colnames(sample_info) |>
  str_replace_all("`", "") |>
  str_replace_all(" ", "_")

sample_info$pct.change <- sample_info$pct.change |>
  str_replace_all("-", "_") |>
  as.factor()

sample_info$pct.change <- relevel(sample_info$pct.change, ref = "0")

TestModels <- function(model, para = FALSE){
# Create a DESeqDataSet
dds <- DESeqDataSetFromMatrix(
  countData = counts.small,
  colData = sample_info,
  design = model
)

# Run the DESeq pipeline
dds <- DESeq(dds, parallel = para)

# Run the differential expression analysis
resultsNames(dds)
genes <- lapply(pct.use, function(x)
{results(dds, contrast = c("pct.change", "0", x)) |> as.data.frame()})
return(genes)
}

pct.use <- levels(sample_info$pct.change)[-1]

#start <- Sys.time()
#de.lfcse <- lapply(models, function(x){TestModels(x)})  
#single.time <- Sys.time() - start

start <- Sys.time()
deseq.results <- lapply(models, function(x){TestModels(x, para = F)})  
multi.time <- Sys.time() - start

#save(deseq.res, file = "jensen/results/benchmarking/deseq/08112023_2000g")

```



```{r get deseq data, fig.width= 11, fig.height= 6}

deseq.res <- deseq.results
# Iterate over the names of de.lfcse
for (sublist_name in names(deseq.res)) {
  # Get the current sublist
  sublist <- deseq.res[[sublist_name]]
  
  # Rename the elements in the sublist using pct.use
  names(sublist) <- pct.use
  
  # Update the sublist in de.lfcse
  deseq.res[[sublist_name]] <- sublist
}

deseq.res.test <- map(deseq.res, ~ set_names(.x, pct.use))


# Function to collect values from nested list DESeq output
calculate_stats <- function(data, column_name, threshold) {
  result <- data %>%
    enframe(name = "sublist_name", value = "data") %>%
    mutate(data = map(data, ~.x %>% enframe(name = "sub_sublist_name", value = "values"))) %>%
    unnest(data) %>%
    mutate(
      mean_value = map_dbl(values, ~mean(.x[[column_name]], na.rm = TRUE)),
      std_dev = map_dbl(values, ~sd(.x[[column_name]], na.rm = TRUE)),
      false_pos = map_dbl(values, ~sum(abs(.x[[column_name]]) < threshold & rownames(.x) %in% stable_genes, na.rm = TRUE)),
      true_pos = map_dbl(values, ~sum(.x[[column_name]] < threshold & rownames(.x) %in% diff_exp_genes, na.rm = TRUE))
    ) %>%
    select(-values)

  return(result)
}

result <- calculate_stats(deseq.res, "padj", 0.05)


result$sub_sublist_name <- result$sub_sublist_name |>
                     str_replace("_","-") |>
                     as.numeric()

result <- result |> 
      mutate(ModelType = factor(case_when(
      str_detect(sublist_name, "unadjusted") ~ "No adjustment",
      str_detect(sublist_name, "raw") ~ "Raw proportions",
      str_detect(sublist_name, "clr") ~ "Centered log ratios",
      str_detect(sub_sublist_name, "pc") ~ "PCA")))

```

```{r plot DESeq output, fig.width=12, fig.height= 10, warning=F}

pal <- c("#fa9fb5", "#f768a1", "#c51b8a", "#74c476", "#31a354", "#a6bddb", "#74a9cf", "#2b8cbe", "#969696")
# Modify legend text 

models.legend <- list(
  unadjusted    = ~ "Group",
  raw_cm        = ~ "Group + CM prop",
  raw_cm_fb     = ~ "Group + CM prop + FB prop",
  raw_cm_fb_im  = ~ "Group + CM prop + FB prop + EC prop",
  clr_cm        = ~ "Group + CLR of CMs",
  clr_cm_fb     = ~ "Group + CLR of CMs + CLR of FBs",
  clr_cm_fb_im  = ~ "Group + CLR of CMs + CLR of FBs + CLR of ECs",
  pc1           = ~ "Group + PC1",
  pc2           = ~ "Group + PC1 + PC2"
  )

p.fp <- result |>
ggplot(aes(x = sub_sublist_name, y = false_pos +0.01, color = sublist_name)) +
  geom_point(alpha = 0.4) + # smaller and more transparent points
  geom_line() +
  #scale_y_continuous(trans = "log2") +
  scale_color_manual(values = pal,
                     name = "Model design",
                     labels = models.legend) +
  theme_minimal() +
  theme(
    legend.position = "none",
    legend.title = element_blank(),
    text = element_text(size = 12),
    panel.grid.major = element_blank(), # remove major gridlines
    panel.grid.minor = element_blank()  # remove minor gridlines
  ) +
  #geom_magnify(from = c(-0.07, 0.07,-3,30), to = c(-0.08,0.08, 380,600), 
  #                 shape = "rect", shadow = F)  +
  labs(y= "False Positives per 200 genes", x = "Simulated fibroblast proportion difference") +
  ggtitle("False positive rate in simulated dataset")

p.tp <- result |>
ggplot(aes(x = sub_sublist_name, y = true_pos + 0.01, color = sublist_name, group = sublist_name)) +
  geom_point(alpha = 0.4) + # smaller and more transparent points
  geom_line() +
  #scale_y_continuous(trans = "log2") +
  scale_color_manual(values = pal,
                     name = "Model design",
                     labels = models.legend) +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    legend.title = element_blank(),
    text = element_text(size = 12),
    panel.grid.major = element_blank(), # remove major gridlines
    panel.grid.minor = element_blank()  # remove minor gridlines
  ) +
  #geom_magnify(from = c(-0.07, 0.07,-3,30), to = c(-0.08,0.08, 380,600), 
  #                 shape = "rect", shadow = F)  +
  labs(y= "True Positives out of 40 DE genes", x = "Simulated fibroblast proportion difference") +
  ggtitle("True positive rate in DE analysis")

design <- c("A
             B")
wrap_plots(A = p.fp, b = p.tp, design = design)

```

