---
title: "Oral Qual plots"
author: "Brian Gural"
date: "2023-10-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 10, message = F, warning = F, error = T)
```



```{r load libs}
libs <- c("tidyverse", "Seurat", "SeuratDisk", "SCpubr", "Biobase", "MuSiC", 
          "reshape2", "SingleCellExperiment") # list libraries here
lapply(libs, require, character.only = T)
rm(libs)
```

```{r load data}
#sn <- LoadH5Seurat("data/processed/single_cell/merges/rau_patterson/09132023/doublets_below_3.h5seurat")
# Data from all sources
sn <- LoadH5Seurat("data/processed/single_cell/merged_old.h5seurat")

# Load bulk data
bulk <- read.csv("data/processed/bulk/cc/06222022/counts_summed_transcripts.csv", row.names = 1)
colnames(bulk) <- paste0("S", extract_numeric(colnames(bulk)))


# Load fractions data
fractions <- read.csv("data/processed/bulk/cc/fractions/jensen_rau_froese_cpm", row.names = 1)
fractions.pheno <- read.csv("data/processed/bulk/cc/fractions/jensen_rau_froese_pheno", row.names = 1)
frac.keep <- fractions.pheno |>   subset(origin != "Jensen") |>   pull(id)
fractions <- fractions[,frac.keep]
fractions.pheno <- fractions.pheno |> subset(origin != "Jensen")
rm(frac.keep)
colnames(fractions.pheno)[1] <- "Sub"

#Load just Rau, raw counts
fractions <- read.csv("data/raw/rau_frac/celltype_counts.csv", row.names = 1)
fractions.pheno <- read.csv("data/raw/rau_frac/celltype_pheno.csv")
colnames(fractions.pheno)[4] <- "Sub"
```

```{r merge bulks}
common.genes <- row.names(fractions)[row.names(fractions) %in% row.names(bulk)]

fractions <- fractions[common.genes,]
#fractions <- 2^fractions
 
bulk <- bulk[common.genes,]

 # Calculate the library size for each sample
lib_sizes <- matrixStats::colSums2(as.matrix(bulk))
cpm <- t(apply(bulk, 1, function(x) (x / lib_sizes) * 1e6)) 
cpm <- cpm + 0.0000001


# Calculate the library size for each sample


bulk.all <- cbind(bulk, fractions)

```

```{r recluster seurat, fig.height=10, fig.width=12}

# Determine percent of variation associated with each PC
pct <- sn[["pca"]]@stdev / sum(sn[["pca"]]@stdev) * 100
# Calculate cumulative percents for each PC
cumu <- cumsum(pct)
# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
co1 <- which(cumu > 90 & pct < 5)[1]
# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = T)[1] + 1
pcs <- max(co1, co2)
    
sn.test <-  sn |> 
FindNeighbors(dims = 1:pcs, reduction = "harmony", verbose = F) |>
FindClusters(resolution = 0.4, verbose = F) |>
RunUMAP(dims = 1:pcs, reduction = "harmony", verbose = F)

do_DimPlot(sn.test, label = T) + NoLegend()

# See if we can  find myofibroblasts
a <- do_FeaturePlot(sn.test, "Postn")
b <- do_FeaturePlot(sn.test, "Acta2")

wrap_plots(a,b)

# And cFBs at homeostasis
c <- do_FeaturePlot(sn.test, "Col1a1")
d <- do_FeaturePlot(sn.test, "Pdgfra")

wrap_plots(a,b,c,d, ncol = 2)
```

```{r find markers}
# Update sn with annotations
sn.mark <- sn.test |>
      #subset(idents = c(0,1,2,3,4,5)) |> 
      subset(features = rownames(sn)[rownames(sn) %in% rownames(bulk.all)])

sn.sce <- sn.mark |> 
  as.SingleCellExperiment(assay = "RNA") |>
  as("SummarizedExperiment")

# Add gene names
rownames(sn.sce) <- rownames(sn.mark@assays$RNA@counts)

# Get markers based on expression ratios between clusters and 1vAll comparisons

scn.markers  <- scran::findMarkers(sn.sce, groups = Idents(sn.mark), pval.type = "some", assay.type = "counts")


.getMarkers <- function(type){
marker <- scn.markers@listData[[type]] |>
  as.data.frame() |>
  #arrange(desc(summary.logFC)) |>
  #slice_head(n = 80) |>
  arrange(FDR) |>
  slice_head(n = 30) |>
  dplyr::select(p.value, FDR, summary.logFC) |>
  mutate(celltype = type) |> 
  rownames_to_column(var = "gene")
return(marker)}

# Collapse list of data frames
all.markers <- lapply(levels(Idents(sn.mark)), function(x){.getMarkers(x)}) |>
               purrr::reduce(full_join)

```

```{r define general clusters}
ecs <- c(0,13, 14)
cardiomyocytes <- c(1,9)
mf <- c(2,7)
vSMC <- 4
profibrotic_FBs <- 5
antigen_presenting_MF <- c(6,10, 18)
m1_mf <- c(8,11)
vascular_ec <-12
myofibroblasts <- 15
nonspecific_fb <- 16
monocytes <- 17

na <- 19
angiogenic_ec <- 20
cell.types <- rep(NA, length(unique(Idents(sn.test))))

cell.types[c(0)+1] <- "Endothelial Cells" 
cell.types[c(12, 13, 14)+1] <- "Endothelial Cells_2" 
cell.types[c(1,9)+1]  <- "Cardiomyocytes"
cell.types[c(3)+1]  <- "FBs"
cell.types[c(5,4, 15, 16)+1]  <- "Pro-Fibrotic FBs"
cell.types[c(2,6,7,8,10,11, 17, 18)+1] <- "MFs"
#cell.types[c(17)+1] <- "Monocytes"
#cell.types[c(4)+1] <- "VSMC"
#cell.types[c(8,11)+1] <- "M1 MFs"
#cell.types[c(15)+1] <- "Myofibroblasts"

sn.clust <- sn.test 
names(cell.types) <- levels(sn.clust)
sn.clust <- RenameIdents(sn.clust, cell.types)


```

```{r redo markers}
# Update sn with annotations
sn.mark <- sn.clust |>
      #subset(idents = c(0,1,2,3,4,5)) |> 
      subset(features = rownames(sn.clust)[rownames(sn.clust) %in% rownames(bulk.all)])

sn.sce <- sn.mark |> 
  as.SingleCellExperiment(assay = "RNA") |>
  as("SummarizedExperiment")

# Add gene names
rownames(sn.sce) <- rownames(sn.mark@assays$RNA@counts)

# Get markers based on expression ratios between clusters and 1vAll comparisons

scn.markers  <- scran::findMarkers(sn.sce, groups = Idents(sn.mark), pval.type = "all", assay.type = "counts")


.getMarkers <- function(type){
marker <- scn.markers@listData[[type]] |>
  as.data.frame() |>
  arrange(desc(summary.logFC)) |>
  slice_head(n = 30) |>
  arrange(FDR) |>
  slice_head(n = 15) |>
  dplyr::select(p.value, FDR, summary.logFC) |>
  mutate(celltype = type) |> 
  rownames_to_column(var = "gene")
return(marker)}

# Collapse list of data frames
all.markers <- lapply(levels(Idents(sn.mark)), function(x){.getMarkers(x)}) |>
               purrr::reduce(full_join)
do_DimPlot(sn.mark)
```

```{r music decon}

bulk.es.exp <- bulk.all[all.markers$gene,] |> 
  sapply(as.integer) %>%
  ExpressionSet(assayData = .) |> 
  exprs()
row.names(bulk.es.exp) <- all.markers$gene


# Run MuSiC with subset genes

# Convert to SingleCellExperiment
sce <- as.SingleCellExperiment(sn.mark, assay = "RNA")
  
# Exclude specified clusters
cells <- levels(Idents(sn.mark))
cells <- cells[!(cells %in% c("unclear")) & !is.na(cells)]
  
# Use MuSiC to estimate cell type proportions
decon <- music_prop(bulk.mtx = bulk.es.exp, sc.sce = sce, markers = all.markers$gene,
                      clusters = "ident", samples = "orig.ident",
                      select.ct = cells, ct.cov = T)
  
# Turn MuSiC output into graph-friendly dataframe
decon.melt <- reshape2::melt(decon$Est.prop.weighted)
colnames(decon.melt) = c('Sub', 'CellType', 'Prop')

rm(sce)
rm(cells)
```

```{r bisque}
library(BisqueRNA)
# Remake seurat as sn.sce without any NA in the idents
sn.sce <- sn.mark |>
          subset(idents = unique(Idents(sn.mark))[!is.na(unique(Idents(sn.mark)))]) |>
          as.SingleCellExperiment(assay = "RNA") 

# Create expression set 
exp_set_sce <- ExpressionSet(
    assayData = as.matrix(assays(sn.sce)$counts),
    phenoData = AnnotatedDataFrame(
        as.data.frame(colData(sn.sce))[c("ident", "orig.ident", "origin")]
    )
)

# Check for nuclei w/ zero expression in marker genes
exp_set_sce <- exp_set_sce[all.markers$gene, ]
zero_cell_filter <- colSums(exprs(exp_set_sce)) != 0
message("Exclude ", sum(!zero_cell_filter), " cells")
exp_set_sce <- exp_set_sce[, zero_cell_filter]

# Make bulk RNAseq into expression set
# exclude low expressed genes 

bulk.qc <- bulk.all[rowSums(bulk.all) > 1, ]
bulk.es <- ExpressionSet(assayData = as.matrix(bulk.qc[all.markers$gene,]))

# Run Bisque
est_prop <- ReferenceBasedDecomposition(
    bulk.eset = bulk.es,
    sc.eset = exp_set_sce,
    cell.types = "ident",
    subject.names = "orig.ident",
    use.overlap = FALSE
)

est_prop$bulk.props <- t(est_prop$bulk.props)

props.bisque <- est_prop$Est.prop.long <- est_prop$bulk.props %>%
    as.data.frame() %>%
    tibble::rownames_to_column("Sample") %>%
    pivot_longer(!Sample, names_to = "cell_type", values_to = "prop")

# Format columns and add meta data for decon method
props.bisque$version <- "bisque"
colnames(props.bisque) <- c('Sub', 'CellType', 'Prop', "version")
```


```{r reformat data}
# Divide fraction and whole data
#decon.melt$version <- "music" 
#decon.melt <- rbind(decon.melt, props.bisque)
decon.frac <- decon.melt|> 
              subset(Sub %in% fractions.pheno$Sub) |> 
              merge(fractions.pheno)


decon.whole <- decon.melt|> 
              #mutate(Sub = paste0("S", Sub)) |> 
              subset(Sub %in% phenos$Sub)
# Unmelt for clustering 
decon.wide <- decon.whole  |> 
              pivot_wider(names_from = "Sub", values_from = "Prop") |> 
              column_to_rownames("CellType") %>%
              mutate_all(as.numeric)

# Add strain + treatment info
decon.whole <- decon.whole |> 
              left_join(phenos)

# Reorder the factor levels of CellType so that they are ordered by sum of their values across groups
decon.melt$CellType <- reorder(decon.melt$CellType, -decon.melt$Prop, FUN = sum)

```

```{r plot fractions, fig.width=10, fig.height = 4}
decon.frac |>
  ggplot(aes(x=Sub, y=Prop, fill=CellType))  +
   geom_bar(stat='identity',
           position = "fill",
           width = 0.8,
           color = "black")+
   scale_color_brewer(palette = "Dark2")  +
   facet_wrap(~cell.type,
              scales = "free_x")+
   ylab("Proportion") +
   theme_minimal() +
   theme(axis.text.x = element_blank(),
        axis.text.y = element_text(color = "black", size = 12),
        axis.ticks = element_blank(),
        strip.text = element_text(size = 15),
        title = element_text(size = 17),
        legend.text = element_text(size = 15),
        panel.background = element_blank(),
        panel.grid = element_blank()) +
   xlab("Sample replicates") 


```

```{r comp plots, echo=FALSE, fig.width=8, fig.height = 14}

my_palette <- c('#78c679', '#d95f0e')

ordered_df <- decon.whole %>%
  filter(CellType == "Cardiomyocytes") %>%  # Filter rows where CellType is "Cardiomyocytes"
  arrange(Prop) %>%  # Sort these rows by Prop in descending order
  pull(Sub)  # Extract the Sub values in this order
# Use the ordered Sub values to reorder the original data frame
decon.whole <- decon.whole %>%
  mutate(Sub = factor(Sub, levels = ordered_df)) %>%
  arrange(Sub)
decon.whole$CellType_wrap = str_wrap(decon.whole$CellType, width = 12)

design <- "AA
           BB"

# Generate boxplot
comp_celltype <- decon.whole   |> 
    filter(Prop > 0.01) %>%
    ggplot(aes(x = as.factor(CellType_wrap), y = Prop, fill = Treatment)) +
    geom_boxplot(position = position_dodge(0.9), width = 0.9, color = "black") +
    geom_jitter(inherit.aes = T, 
                position = position_jitterdodge(jitter.width = 0.1, dodge.width = 0.9),
                size = 2) +
    theme(axis.text.x = element_text(color = "black", size = 20, angle = 30),
          axis.text.y = element_text(color = "black", size = 20),
          axis.ticks = element_blank(),
          legend.position = c(0.9, 0.9),
          legend.justification = c("right", "top"),
          legend.box.just = "right",
          legend.margin = margin(6, 6, 6, 6),
          strip.text = element_text(size = 20),
          title = element_text(size = 30),
          legend.text = element_text(size = 25),
          panel.background = element_rect(fill='transparent'),
          plot.background = element_rect(fill='transparent', color=NA),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          legend.background = element_rect(fill='transparent'),
          legend.box.background = element_rect(fill='transparent'),
          axis.title.x = element_blank(),
          plot.margin = margin(5, 5, 0, 5),
          axis.title.y = element_text(size = 25)) +
    labs(y = "Proportion", 
         fill = "Treatment") +
    scale_fill_manual(values = my_palette) 

# Plot each sample in a barchart stacked by cell type
comp_sample <- decon.whole  |>
    subset(Prop > 0.01) |> 
    ggplot(aes(x=Sub, y=Prop, fill=CellType_wrap))  +
    geom_bar(stat='identity',
           position = "fill",
           width = 1,
           color = "black")+
    scale_fill_brewer(name = "Cell Type",
                      palette = "Set2") +
    theme_minimal() +
    theme(strip.text = element_text(size = 20),
          title = element_text(size = 20),
          axis.text.y  = element_text(size = 20),
          legend.text = element_text(size = 18),
          axis.ticks.x = element_blank(),
          panel.grid = element_blank(),
          legend.position = 'bottom',
          legend.justification = 'left',
          legend.margin = margin(t = -5, r = 0, b = 0, l = 0, unit = "pt"),
          axis.text.x = element_blank(),
          plot.margin = margin(0, 5, 1, 5),
          plot.caption = element_text(size = 22, hjust = -0.3, face = "bold"),
          plot.tag = element_text(size = 22, hjust = -0.3),
          axis.title.y = element_text(size = 25)) +
    labs(y = "Proportion", 
         x = "Samples") +
    guides(fill = guide_legend(nrow = 2)) 
  


wrap_plots(A = comp_celltype, B = comp_sample, design = design)


```




```{r clr transform}
# use clr or ilr to incorp the compositions into the design matrix
comps.clr <- compositions::clr(t(decon.wide))

colnames(comps.clr) <- paste0("clr.", colnames(comps.clr))

```

```{r deseq pre, fig.width = 12, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}

# Prepare for DESeq2
count_data <- bulk[,-1]

bulk <- mutate_all(bulk, function(x) round(as.numeric(as.character(x)), digits = 0)) # round to integers

pheno.reorder <- phenos |> mutate(Sub = factor(Sub, levels = colnames(bulk))) |> arrange(Sub)
# Prepare sample information
sample_info <- data.frame(
  row.names = colnames(bulk),
  genotype = as.factor( pheno.reorder$Sample.ID),
  treatment = as.factor(pheno.reorder$Treatment)
)


sample_info$treatment <- relevel(sample_info$treatment, ref = "Ctrl")

```


```{r deseq clr}
library(DESeq2)
sample.clr <- cbind(sample_info, comps.clr)

# Create a DESeqDataSet
dds.clr <- DESeqDataSetFromMatrix(
  countData = bulk,
  colData = sample.clr,
  design = ~ genotype + treatment + genotype:treatment + clr.Cardiomyocytes +  clr.FBs #+ clr.endothelial_cells
)

# Run the DESeq pipeline
dds.clr <- DESeq(dds.clr)

# Run the differential expression analysis
res.clr <- results(dds.clr, contrast = c("treatment", "Iso", "Ctrl"))

```

```{r deseq raw}
# Create a DESeqDataSet
dds.raw <- DESeqDataSetFromMatrix(
  countData = bulk,
  colData = sample_info,
  design = ~ genotype + treatment + genotype:treatment
)

# Run the DESeq pipeline
dds.raw <- DESeq(dds.raw)

# Run the differential expression analysis
res.raw <- results(dds.raw, contrast = c("treatment", "Iso", "Ctrl"))
```


```{r compare deseq clr, fig.width = 8, fig.height= 6, echo=TRUE, message=TRUE, warning=FALSE}
# Compare results
comparison.clr <- merge(as.data.frame(res.clr), as.data.frame(res.raw), by = "row.names", suffixes = c(".clr", ".raw"))

# Calculate difference in -log10 p-values
comparison.clr$p_diff <- -log10(comparison.clr$padj.clr) - -log10(comparison.clr$padj.raw)

# Make categorical column
comparison.clr <- comparison.clr |>
        mutate(clr.sig = ifelse(padj.clr <= 0.05 & padj.raw >= 0.05, T, F),
               raw.sig = ifelse(padj.clr >= 0.05 & padj.raw <= 0.05, T, F),
               both.sig = ifelse(padj.clr <= 0.05 & padj.raw <= 0.05, T, F)) 

# Get the gene names for the top 5 genes by change in adjusted p value, within lost and gained sig groups
gained.clr <- comparison.clr |>  
  subset(clr.sig == T) |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

lost.clr <- comparison.clr |>  
  subset(raw.sig == T) |> 
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 5)

top.clr <- comparison.clr |> 
  arrange(padj.clr) |>
  slice_head(n = 15)

top.lost.raw <- comparison.clr |>
  subset(raw.sig == T) |> 
  arrange(padj.raw) |>
  slice_head(n = 10)
```

```{r plot clr, fig.width = 14, fig.height= 12}
library(ggrepel)

limit <- comparison.clr |> 
  na.omit() |>
  arrange(desc(abs(p_diff))) |>
  slice_head(n = 1) |>
  pull(p_diff) * c(1, -1)

plot.clr <- ggplot(comparison.clr, aes(x = log2FoldChange.clr, y = -log10(padj.clr), color = p_diff)) +
  geom_point(alpha = 0.8, size = 4) +
  scale_color_distiller(type = "div")+
  geom_text_repel(data = gained.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 40, segment.linetype = 2, segment.size = 0.6, color = "orange", force_pull = 0.01, min.segment.length = 0) + # label most increase sig genes 
  geom_text_repel(data = lost.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "purple", force_pull = 0.01, min.segment.length = 0) + # label most decrease sig genes 
  geom_text_repel(data = top.clr, aes(label = Row.names), size = 8, box.padding = unit(0.35, "lines"), 
                  force = 20, segment.linetype = 2, segment.size = 0.6, color = "black", force_pull = 0.01, min.segment.length = 0) + # label most sig genes 
  theme_minimal() +
  labs(x = "log2(Fold Change)", y = "-log10(adjusted p-value)", color = "Change in -log10(p-value)") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_vline(xintercept = 0, linetype = "solid") + # add a line for p=0.05
  theme(legend.position = "bottom",
        axis.text.x = element_text(color = "black", size = 25),
        axis.text.y = element_text(color = "black", size = 25),
        axis.ticks = element_blank(),
        title = element_text(size = 25),
        legend.text = element_text(size = 25),
        legend.key.size = unit(0.7, 'in'),
        legend.title = element_text(vjust = 0.7),
        axis.title.y = element_text(color = "black", size = 25),
        axis.title.x = element_text(color = "black", size = 25),
        panel.background = element_rect(fill='transparent'),
        plot.background = element_rect(fill='transparent', color=NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  ggtitle("DE w/ clr-based composition in model", subtitle = "Colored text show genes that became or lost significance, top 5 by change")# move legend to the bottom

plot.clr
```

```{r immune + FB markers, fig.width = 30, fig.height= 10, echo=FALSE, message=FALSE, warning=FALSE}

# Screening 

plotGene <- function(seurat, slot = "scale.data", gene, plot.title = T, plot.legend = T){
  
  seurat |>
    do_FeaturePlot(gene, reduction = "umap", slot = slot, 
                   legend.position = if(plot.legend == T){"bottom"}else{"none"}, 
                   legend.width = 2, legend.length = 25,  order = T) +
    theme(title = element_text(size = 22),
          legend.title = element_blank()) +
    labs(title = if(plot.title == T){paste0(gene)}else{" "})
}

genes <- c("Ttn", "Col3a1", "Lancl3", "Errfi1")



# plot new
plots <- c()
plots[["cell"]] <- do_DimPlot(sn, reduction = "umap", label = T, repel = T)  +  
  NoLegend() + theme(title = element_text(size = 22))
plots.exp <- lapply(genes, function(x){plotGene(seurat = sn, slot = "data", gene = x, plot.legend = F)})

plots.exp <- c(plots, plots.exp)
wrap_plots(plots.exp, 
           ncol = 5)
```